### PR TITLE
feat: on-demand export of buffered traffic recording + ring-buffer bug hunt + JMH hardening

### LINKED ISSUE #1282: Traffic recording: on-demand export of the disk ring buffer contents, plus bug-hunt & JMH performance hardening
## Summary

Add a single **on-demand export** operation to the traffic-recording subsystem that safely exports the
**current valid contents of the disk ring buffer** as a downloadable zip, so recorded traffic can be
replayed/analysed on another machine (developer box, regression-testing harness). Also **harden** the
subsystem with a dedicated bug hunt and a JMH performance suite plus the optimizations those uncover.

The export is **catalog-specific**, triggered over the **gRPC traffic-recording service**, and produces a zip
consumed by evitaLab / gRPC-web — it is **not** exposed through the Java `EvitaClient` driver (see Scope).

Central engineering hazard called out explicitly: the disk ring buffer keeps being written and rotated
(`head`/`tail` advance, oldest sessions evicted) **while the export runs** — the export must take a
**consistent snapshot of the valid window under concurrent mutation** and must not corrupt or stall the live
recorder, nor emit truncated/garbled records.

---

## Background — how traffic recording works today

(Grounding for whoever picks this up; file references are approximate anchors, not exact lines.)

- **Rich recorder.** When recording is active, `TrafficRecordingEngine` runs the `ServiceLoader`-provided
  `OffHeapTrafficRecorder` (`evita_store/evita_traffic_engine/.../store/traffic/OffHeapTrafficRecorder.java`).
  When inactive it uses `NoOpTrafficRecorder` (no buffers, records nothing).
- **The disk ring buffer always exists whenever recording is active.** `OffHeapTrafficRecorder.init()`
  unconditionally allocates a small (~4 MiB default) off-heap block-staging area **and** a `DiskRingBuffer`
  (`.../store/traffic/DiskRingBuffer.java`, `trafficDiskBufferSizeInBytes` = **32 MiB** default). There is no
  "disk-disabled" mode. The disk file is **transient** (recreated empty on construction; pre-allocated to full
  size; not durable across restart).
- **Memory is transient staging; the disk ring buffer is the retained store.** On `closeSession`, a finalized
  session is queued; a background task (`freeMemory()`, paced by `trafficFlushIntervalInMilliseconds`, default
  **60 s**; `0` = immediate) drains it into the `DiskRingBuffer` (advancing `ringBufferTail`, evicting the
  oldest and advancing `ringBufferHead`), then frees the memory blocks. The off-heap staging fully drains into
  the disk buffer within the flush interval.
- **Read-back is disk-only.** `getRecordings` / `getRecordingsReversed` read the disk ring buffer via a lazily
  built in-memory `TrafficRecordingIndex` (`IndexNotReady` if not yet built). Concurrency between writer and
  readers is coordinated with OS `FileChannel` range locks (exclusive write / shared read) plus per-record
  re-validation that a session's bytes are still inside `[head, tail]`; a record overwritten mid-read is
  dropped. Sessions still in memory staging (open, or finalized-but-not-yet-drained) are not visible.
- **A live-capture export already exists** — `TrafficRecorderTask` + its nested `ExportSessionSink`
  (`evita_engine/.../core/traffic/task/TrafficRecorderTask.java`), triggered by `StartTrafficRecording(exportFile=true)`.
  It streams raw session bytes into a chunked zip (`traffic_recording_<seq>.bin` entries + `metadata.txt`) as
  sessions are **evicted** from the ring buffer, via the `RandomAccessFileSessionSink` callback. It is
  **forward-only** (captures from the moment recording starts). Exported zips are read back by
  `InputStreamTrafficRecordReader` (`.../store/traffic/InputStreamTrafficRecordReader.java`).

**The gap this issue fills.** There is **no on-demand export of what is *already* in the disk ring buffer.**
Today you can only capture *new* traffic going forward via the live task. This issue adds "export the current
buffer window right now."

### Why a single disk-buffer export (scope decision)

The disk ring buffer always exists whenever recording is active and *is* the retained store; the ~4 MiB memory
staging fully drains into it within ~60 s. Therefore a "memory-only" export is a marginal, ephemeral sliver
(and cannot cleanly include in-flight *open* sessions, which are incomplete), and "memory + disk" is
effectively just "the disk buffer." Collapsing to a single **disk-ring-buffer export** removes the complex
"snapshot the off-heap staging under mutation" path entirely and keeps the head/tail consistency concern only
where it is literal — on the disk span.

---

### R1. Single catalog-specific export operation

- **`exportTrafficRecording()`** — export the **current valid contents of the disk ring buffer** (the sessions
  presently retained in the `[head, tail]` window) to a downloadable **zip**.
- Returns a background task (`Task`/`ServerTask`) whose result is a `FileForFetch`, following the existing
  export-producer pattern (`BackupTask` / `TrafficRecorderTask` →
  `ExportService.storeFile(name+".zip", description, "application/zip", origin)` → `FileForFetch`).
- **Optional flush-before-export** (implementation decision): the export MAY first trigger `freeMemory()` to
  drain finalized-but-not-yet-flushed sessions into the disk buffer so very recent closed sessions are
  included. Caveat to weigh: force-draining advances the ring and can evict the oldest sessions to make room —
  i.e. it slightly shifts the retained window. If this trade-off is unwanted, export the disk window as-is and
  accept that the last ≤ flush-interval of closed sessions may be absent. Either way, **open (in-flight)
  sessions are out of scope** — they are incomplete and must not be exported.

### R2. Safe, consistent snapshot of *valid* contents under concurrent mutation

The disk buffer keeps being written and rotated during export. The export must:

- Capture a **stable snapshot** of the set of valid sessions / the `[head, tail]` window at export start, and
  **reuse the existing consistency machinery** rather than inventing a new one: the reader already takes a
  shared `FileChannel` range lock and re-validates each session against the live `[head, tail]` before
  emitting it (`isSessionLocationStillInValidArea`, `lockAndRead` in `DiskRingBuffer`).
- Only export **valid, complete** sessions. A session that is evicted / overwritten by ring rotation
  mid-export must be **skipped cleanly** — never a truncated or garbled record in the zip.
- **Count and report** every skipped/overwritten session in the export `metadata.txt` (and ideally a metric).
  A silently short export is a correctness bug for regression testing (see H1).
- Must **never** stall or corrupt the live recorder, nor block request threads for the duration of the export.

The natural implementation mirrors `ExportSessionSink.compressToFinalDestination(...)` but **pull-driven** over
the current window instead of eviction-driven: seek the disk read handle to each valid session's start, copy
its raw `<lead descriptor><payload>` bytes into the zip under a shared lock, re-validating as it goes.

### R3. Format must be re-consumable downstream

The whole point is "analyse on another machine / use in regression testing," so the exported zip **must be
readable back** by the existing `InputStreamTrafficRecordReader` — i.e. emit the **same `.bin` / Kryo chunk
format** the live `ExportSessionSink` already produces (chunked by `exportFileChunkSizeInBytes`, default
16 MiB, plus a `metadata.txt` entry). Do not invent a second, unreadable format. If the reader needs a small
extension to consume the new artifact, extend it and cover it with a round-trip test. Add an explicit
**export → `InputStreamTrafficRecordReader` round-trip test** as an acceptance gate.

### R4. gRPC surface (catalog specific)

- Add **one** RPC to the **session-bound** service `GrpcEvitaTrafficRecordingService`
  (`.../grpc/GrpcEvitaTrafficRecordingAPI.proto`) — catalog-specificity comes from the bound session
  (`ServerSessionInterceptor` → `session → catalog.getTrafficRecordingEngine()`), same as the existing
  start/stop RPCs. Insert after `StopTrafficRecording`.
- Reuse the existing `GetTrafficRecordingStatusResponse` (wraps `GrpcTaskStatus`) as the response type — the
  RPC kicks off the export task and returns its task status.
- Implement the `@Override` method in `EvitaTrafficRecordingService`
  (`.../grpc/services/EvitaTrafficRecordingService.java`) mirroring `startTrafficRecording`: run inside
  `executeWithClientContext(session -> ...)`, call the new session/engine export method, return the task
  status. No `GrpcProviderRegistrar` change needed (service already registered).
- **Download stays on the server-wide management API** — the produced zip is discovered/downloaded via the
  existing `EvitaManagementService.ListFilesToFetch` (filter by `origin`) + streaming `FetchFile`. No new
  download path.
- Unlike `Start/StopTrafficRecording`, the on-demand export is a plain per-catalog read and must **not** be
  gated by the server-wide `TrafficRecorderTask` singleton guard.

### R5. Engine / task plumbing (the real work — not a thin proto wrapper)

- New method on `TrafficRecordingEngine` to export a consistent snapshot of the disk ring buffer, delegating
  into `OffHeapTrafficRecorder` / `DiskRingBuffer` where the buffer lives. Only the rich recorder supports it;
  `NoOpTrafficRecorder` / a plain (non-reader) recorder must fail with a clear `EvitaInvalidUsageException`
  (mirror the existing `getRecordings` guard).
- New `ServerTask` alongside `TrafficRecorderTask` (`.../core/traffic/task/`), e.g.
  `TrafficRecordingExportTask extends ClientCallableTask<…, FileForFetch>`, that: creates the zip via
  `ExportService.storeFile(...)`, streams the snapshot into it (raw byte copy per R2/R3), reports progress via
  `updateProgress(0..100)`, deletes the partial file on failure (`fileForFetchFuture().getNow(null)` →
  `deleteFile`), and returns the `FileForFetch`. Tag `origin` with the task class name and embed the catalog
  name + timestamp in the file name / description.
- New session-contract method (near `EvitaSession.startRecording`, `.../core/session/EvitaSession.java`) that
  builds and `scheduler.submit(...)`s the export task, obtaining the singleton `ExportService` via
  `evita.management().exportService()`.

---

### H1. Bug hunt (use a strong model to find + reproduce, then TDD the fixes)

Follow the repo's test-first workflow: write a failing test that reproduces each confirmed defect, then fix.
Concrete anchors to investigate (verify each before asserting it's a bug):

- **Export vs. ring rotation race** — an export reading the `[head, tail]` window while the writer advances the
  tail and evicts the head. Confirm sessions overwritten mid-export are handled deterministically (skipped +
  **counted/reported**, never producing a truncated/garbled record in the zip).
- **Silent record drops** — the read path returns `null` on `OverlappingFileLockException` / failed
  `[head, tail]` re-validation. Verify these are surfaced (metric / metadata counter), not silently swallowed —
  a silently short export is a correctness bug for regression testing.
- **`IndexNotReady` races** — first read after startup / after a long idle triggers a lazy index (re)build;
  ensure the export interacts correctly (waits / builds / clear error) instead of failing spuriously.
- **Predicate AND/OR mismatch** — `TrafficRecorder.createRequestPredicate` combines criteria with **AND**,
  while `TrafficRecordingReader`'s javadoc says OR. Reconcile the doc and the behavior; add tests pinning the
  intended semantics.
- **Memory-shortage / block-exhaustion paths** — `MemoryNotAvailableException` → session self-finishes as
  `MEMORY_SHORTAGE`; verify no block leak, no partial session written to disk, correct free-block return.
- **Flush + export interaction** — if the export triggers `freeMemory()`, verify it does not deadlock/contend
  with the periodic flush task or a concurrently attached live-export `SessionSink`, and that eviction caused
  by the forced drain is consistent.
- **Reverse-read correctness** under wrap and eviction.
- Broaden coverage of `OffHeapTrafficRecorder`, `DiskRingBuffer`, `SessionTraffic`, and the serializers to the
  project's ≥70% line-coverage bar for the touched classes.

### H2. JMH performance suite + optimizations

Add JMH benchmarks (per repo JMH conventions; long-running perf tests belong in the perf module, not
`functional_tests`) covering the traffic engine's hot paths, then apply the optimizations they justify.
Concrete anchors:

- **Write path throughput/allocation** under realistic sampling — `recordQuery`/`recordFetch`/`recordMutation`
  → `SessionTraffic` serialization, Kryo pooling, single-writer CAS drain, block allocation + free-block queue
  contention. Measure alloc churn (this subsystem sits on the query hot path).
- **Flush drain cost** — `freeMemory()` batching under different `trafficFlushIntervalInMilliseconds` and load;
  disk `FileChannel` range-lock contention (writer exclusive vs. reader/exporter shared).
- **Index (re)build latency** — cost of building `TrafficRecordingIndex` on first read / after idle, and the
  `IndexNotReady` window.
- **Reverse read** — `getSessionRecordsReversedStream` currently buffers each session into an `ArrayList` and
  reverses it (flagged inefficient). Benchmark and optimize.
- **The new export path** — throughput and interference: measure the live recorder's write latency/throughput
  **while an export runs**, to prove R2's non-interference guarantee, plus the raw-byte-copy throughput of the
  export itself.
- Capture before/after numbers for any optimization in the PR description.

---

## Suggested implementation checklist

- [ ] `TrafficRecordingEngine` — disk-buffer export method delegating to the rich recorder; guard non-reader recorders.
- [ ] `OffHeapTrafficRecorder` / `DiskRingBuffer` — pull-driven consistent-snapshot export of the current `[head, tail]` window (raw `<descriptor><payload>` byte copy under shared lock + re-validation, skipped-record counting); optional pre-export `freeMemory()`.
- [ ] `InputStreamTrafficRecordReader` — verify (extend if needed) it reads the new artifact; add a round-trip test (R3).
- [ ] `TrafficRecordingExportTask` (`.../core/traffic/task/`) — zip via `ExportService.storeFile`, progress, partial-file cleanup, `metadata.txt` incl. exported + skipped-record counts.
- [ ] `EvitaSession` (+ contract) — `exportTrafficRecording(...)` building + submitting the task.
- [ ] Proto `GrpcEvitaTrafficRecordingAPI.proto` — one RPC after `StopTrafficRecording`, reuse `GetTrafficRecordingStatusResponse`.
- [ ] `EvitaTrafficRecordingService` — one `@Override` impl mirroring `startTrafficRecording`.
- [ ] Bug-hunt fixes (H1), each with a failing-first reproduction test.
- [ ] JMH suite (H2) + applied optimizations with before/after numbers.
- [ ] Docs: note the new export operation where traffic recording is documented.

## Acceptance criteria

- The export produces a downloadable zip via `ListFilesToFetch` + `FetchFile`, catalog-specific, triggered over
  `GrpcEvitaTrafficRecordingService`.
- Exported zip round-trips through `InputStreamTrafficRecordReader` (R3 test green).
- Export does not corrupt or stall the live recorder; sessions evicted/overwritten mid-export are skipped and
  **counted/reported**, never producing a truncated record or a silently short file (R2 + H1).
- Bug-hunt findings fixed with regression tests; touched classes ≥70% line coverage.
- JMH suite committed; optimizations applied where justified, with before/after numbers in the PR.

## Scope / non-goals

- **Single export operation only** — export the current disk ring buffer window. No separate memory-only
  export (see "Why a single disk-buffer export"); no off-heap-staging snapshot path; open in-flight sessions
  are excluded.
- **No Java `EvitaClient` driver support.** The traffic-recording gRPC service has no Java client stub today;
  this export RPC will **not** be accessible via the Java driver. It is consumed via gRPC-web / evitaLab / C#.
- Durable/persistent traffic storage across restart is out of scope — the disk ring buffer remains a bounded,
  transient window. Retaining history beyond the buffer size remains the job of the forward-only live-capture
  `SessionSink` export, not this operation.
- No new `ExportService` storage backend — reuse the existing per-engine singleton (files are global, tagged by
  `origin`, catalog name embedded in the file name).

### PR DESCRIPTION
**Pull request by novoj**
## Summary

Adds an **on-demand export** of the currently buffered traffic-recording window (the recent
disk ring-buffer window the server always keeps), independent of the existing start/stop streaming
recorder task. A caller triggers `ExportTrafficRecording` over the gRPC traffic-recording API and
downloads the resulting ZIP through the file-fetch API (`ListFilesToFetch` + `FetchFile`). The
export takes a consistent one-shot snapshot without starting, stopping, or interrupting live
recording.

Along the way this branch hunts and fixes a family of concurrency/wrap correctness bugs in the ring
buffer, and hardens the traffic write/read paths with a JMH suite plus one applied allocation
optimization.

Closes #1282.

## What's new

- **`TrafficRecordingExporter`** (evita_api) — sibling capability interface to `TrafficRecordingReader`,
  guarded by `instanceof` at the engine boundary (no new `evita_engine → evita_traffic_engine`
  dependency; SPI boundary preserved).
- **`DiskRingBuffer.exportSnapshot`** — freezes a snapshot, walks oldest→newest, span-locks each
  session, copies raw bytes wrap-aware, skips (and counts) sessions concurrently overwritten or
  evicted mid-walk. Never emits a partial session.
- **`TrafficRecordingExportTask`** (one-shot callable) + `EvitaSession.exportTrafficRecording` +
  gRPC `ExportTrafficRecording` RPC + `GrpcExportTrafficRecordingRequest` (optional chunk-file size).
- The ZIP contains chunked `traffic_recording_*.bin` files (each fully replayable via
  `InputStreamTrafficRecordReader`) and a `metadata.txt`. **metadata.txt additions** over the
  streaming report: `skipped N sessions (evicted or write-locked during export)` and
  `snapshot contained N sessions`. The generated file carries the **origin `TrafficRecordingExportTask`**
  (vs the streaming report's `TrafficRecorderTask`), so clients listing files can tell the two apart.

## Bug hunt

Each fix carries a failing-first regression test unless noted.

| Defect | Resolution |
|---|---|
| Writer crash (`OverlappingFileLockException`) when a reader holds an overlapping lock | In-JVM span-lock redesign replaces OS `FileLock`; test |
| Reader-vs-reader silent drop on overlapping spans | Shared spans coexist; give-up counter; test |
| Non-volatile `ringBufferHead`/`ringBufferTail` | Made volatile; pinning test |
| `RingBufferInputStream` three wrap bugs (single-byte off-by-one, bulk no-advance, skip no-seek) | Fixed; 5 tests |
| `lockAndWrite` swallowed `IOException` | Fixed as a side effect of the span-lock rewrite; regression test |
| Filter predicate AND-vs-OR doc mismatch | Reconciled JavaDoc to the actual AND-across-criteria / OR-within semantics |
| `ExportSessionSink.compressToFinalDestination` wrap safety (old streaming path) | Probed; pinned behavior |
| Sampling 0/100 contradictions across JavaDoc/code | Reconciled JavaDoc (0 disables, 100 records all) |
| `stopTrafficRecording` sent two `onNext` responses | Removed duplicate; gRPC red→green regression test |
| Block leak on `MemoryNotAvailableException` + wrong `DATA_TOO_LARGE` message | Fixed |
| `readSessionRecords` modulo edge case at exact-wrap boundary | Probed/pinned |
| Partial export file leak on failure | Hoisted handle + `finally` cleanup + S3-exceptional-future guard; test |
| `isSessionLocationStillInValidArea` returns false for a fully-packed buffer (`head == tail`) | Disambiguate via `sessionLocations.isEmpty()`; 2 tests |
| **validity ≠ identity**: an evicted+reused slot passes the validity check → export copies the wrong session → torn `.bin` | Per-record on-disk identity check against the globally-monotonic seq; deterministic rotation-under-export test |

> The last two emerged during the hunt and are not in the issue's enumerated list. The
> validity-vs-identity fix is the headline correctness fix (the rotation-under-export corruption
> class). A stale malformed-FQN SPI registration file was also removed.

## Performance

JMH; AMD Ryzen AI 9 HX 370, OpenJDK 21, `@Fork(1)` `@Warmup(3×1s)` `@Measurement(5×1s)`.

The span-lock redesign (removing 2 syscalls/record on the read path) already carried the win, so the
numbers-driven pass applied only what it justified.

**Applied — `recordQuery` label-merge allocation cut** (gc.alloc.rate.norm, B/op; deterministic,
tight bars):

| method | payload | before | after | Δ |
|---|---:|---:|---:|---:|
| recordQuery (t1) | 64 | 3 639.4 | 2 983.4 | **−656 (−18 %)** |
| recordQuery (t1) | 1024 | 4 615.8 | 3 959.8 | **−656 (−14 %)** |
| recordQuery (t1) | 4096 | 7 795.6 | 7 083.6 | **−712 (−9 %)** |
| recordQuery (t16) | 64 | 3 765.5 | 3 011.4 | **−754** |

**No regression vs the Phase-0 baseline** (AverageTime, µs/op): forward read −10–15 %, index re-build
≈flat, flush −20–35 %.

**Dropped on evidence** (measured, not speculated): reverse-read buffering (rev/fwd ≈ 1.0),
read per-session lock (already faster than baseline), index per-record identity (a first "+26–36 %"
proved to be concurrent-load noise; quiet re-run at baseline levels), `freeBlocks` fair→lock-free
queue (delta inside noise — reverted).

**Rotation-under-export acceptance gate certified** — new `TrafficRecordingExportInterferenceBenchmark`:
writers stay the same order of magnitude with continuous export (~12–33 % at 4 threads), zero writer
exceptions, and the disk ring is confirmed to actually wrap during measurement.

## Acceptance gates (from the issue)

1. ✅ Zip downloadable via `ListFilesToFetch` + `FetchFile`, catalog-specific, triggered over `GrpcEvitaTrafficRecordingService`.
2. ✅ Round-trip test green (`InputStreamTrafficRecordReader` consumes every exported `.bin`).
3. ✅ Rotation-under-export stress green; skipped sessions counted in `metadata.txt`; JMH interference numbers show no writer stall/corruption.
4. ✅ Bug fixes each carry a failing-first regression test; touched classes ≥ 70 % line coverage.
5. ✅ JMH suite committed with before/after numbers for every applied optimization.

## Docs

- `documentation/user/en/operate/observe.md` + `documentation/user/cs/operate/observe.md`:
  new "On-demand traffic recording export" subsection (distinct from the streaming report), noting the
  `metadata.txt` additions and the file origin.
- JavaDoc on all new contract types/methods (`TrafficRecordingExporter`, engine/task/settings).

## Non-goals (scope guard)

Single disk-buffer export only (no memory-only/off-heap staging snapshot, no open in-flight
sessions); no Java `EvitaClient` exposure; no durable cross-restart storage; no new `ExportService`
backend.

### CODE REVIEW

File: evita_store/evita_traffic_engine/src/main/java/io/evitadb/store/traffic/stream/RingBufferInputStream.java [hunk 1]
```diff
@@ -57,45 +57,74 @@ public RingBufferInputStream(
...
 	}
 
 	@Override
 	public long skip(long n) throws IOException {
 		this.position += n;
-		if (this.position > this.inputBufferSize) {
+		if (this.position >= this.inputBufferSize) {
 			this.position = this.position % this.inputBufferSize;
-			this.delegatingInputStream.seek(this.position);
 		}
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_store/evita_traffic_engine/src/main/java/io/evitadb/store/traffic/stream/RingBufferInputStream.java:114]: Fixed in bbb11313f. Both `skip` and `skipNBytes` now short-circuit for non-positive `n` — `skip` returns 0 and `skipNBytes` no-ops, matching the `InputStream` contract. Previously a negative `n` walked `position` backwards past the wrap guard into a negative seek; the only live caller (Kryo) never passed non-positive, so this was latent, but the guard is correct and cheap. Extended it to `skipNBytes` too since it shared the same unguarded `+= n`.

File: evita_engine/src/main/java/io/evitadb/core/traffic/task/TrafficRecordingExportTask.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+			this.outputStream.putNextEntry(new ZipEntry("metadata.txt"));
+			this.outputStream.write("Traffic recording export: \n".getBytes());
+			this.outputStream.write(("\n   - started at " + this.startTime.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)).getBytes());
+			this.outputStream.write(("\n   - finished at " + finishTime.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)).getBytes());
+			this.outputStream.write(("\n   - duration " + StringUtils.formatDuration(Duration.between(this.startTime, finishTime))).getBytes());
+			this.outputStream.write(("\n   - exported " + summary.exportedSessionCount() + " sessions").getBytes());
+			this.outputStream.write(("\n   - exported " + StringUtils.formatByteSize(summary.exportedByteCount()) + " of data").getBytes());
+			this.outputStream.write(("\n   - skipped " + summary.skippedSessionCount() + " sessions (evicted or write-locked during export)").getBytes());
+			this.outputStream.write(("\n   - snapshot contained " + summary.totalSessionCount() + " sessions").getBytes());
+			this.outputStream.closeEntry();
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_engine/src/main/java/io/evitadb/core/traffic/task/TrafficRecordingExportTask.java:205]: Fixed in bbb11313f — `metadata.txt` is now written with an explicit `StandardCharsets.UTF_8`. The content is pure ASCII so the emitted bytes were identical on any real platform, but the explicit charset is the correct, portable form. Applied the same fix to the sibling streaming path (`TrafficRecorderTask`) this sink mirrors, so the two stay consistent.

File: evita_test/evita_functional_tests/src/test/java/io/evitadb/core/traffic/task/TrafficRecordingExportTaskTest.java [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Pins the on-demand {@link TrafficRecordingExportTask#doExport()} contract in isolation, driving the task's
+ * callable directly against a stubbed {@link TrafficRecordingEngine} and a fake {@link ExportService}:
+ *
+ * - a successful export streams every session the engine yields into a zip archive (one `.bin` entry per
+ *   session plus a trailing `metadata.txt`) and returns the finished {@link FileForFetch},
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_test/evita_functional_tests/src/test/java/io/evitadb/core/traffic/task/TrafficRecordingExportTaskTest.java]: Fixed in bbb11313f. Reworded the JavaDoc (and the matching `@DisplayName` and assertion message, which carried the same wrong claim) to describe the actual chunked layout — sessions packed into `traffic_recording_*.bin` entries until the configured chunk size rolls over, not one entry per session.

File: evita_store/evita_traffic_engine/src/main/java/io/evitadb/store/traffic/OffHeapTrafficRecorder.java [hunk 4]
```diff
@@ -590,6 +686,83 @@ protected byte[] create() {
...
+								sessionLocation,
+								finalizedSession.getSessionId(),
+								finalizedSession.getCreated(),
+								finalizedSession.getDurationInMillis(),
+								finalizedSession.getRecordingTypes(),
+								finalizedSession.getLabels(),
+								finalizedSession.getFetchCount(),
+								finalizedSession.getBytesFetchedTotal()
+							);
+						} catch (MemoryNotAvailableException ex) {
```
[the remark this answers is not in the document]
Comment (novoj) [reply evita_store/evita_traffic_engine/src/main/java/io/evitadb/store/traffic/OffHeapTrafficRecorder.java]: Fixed in b79bee0a8. The per-session `try` now catches `MemoryNotAvailableException | UnexpectedIOException`, and the recovery frees the session's *still-held* blocks by re-iterating `getMemoryBlockIds()` and skipping the `freedBlocks` count already offered back inside the loop — so the block that failed and every later block are returned, with no double-free of the ones written before the failure. The session is counted as dropped/missed exactly as the memory-shortage path does. Added a failing-first regression test (`shouldReturnMemoryBlocksWhenDiskWriteFailsMidDrain`) that swaps in a `DiskRingBuffer` spy throwing `UnexpectedIOException` on the 2nd `append()` mid-session and asserts the free pool is fully restored (`totalBlocks`, catching both leak and double-free) and the drop counter incremented once.

**Comment by novoj** [comment]
Thanks for the deep read at HEAD — dispositions on the review-body findings, all in commit b79bee0a8 unless noted:

**MAJOR**
- **#1 drain block leak** — fixed + failing-first regression test (see the inline thread).
- **#2 span-lock writer starvation** — fixed. Went with the writer-preference option: `RingBufferSpanLock` now tracks pending exclusive spans, and `tryAcquireShared` gives up immediately when the candidate conflicts with a pending exclusive. This is consistent with the existing "readers give up immediately" asymmetry, so a steady overlapping read/export stream can no longer starve the waiting writer. The pending marker is cleared in a `finally` (including the interrupted path) so an aborted writer never leaves a phantom span. JavaDoc updated — writer preference now *enforces* the "bounded by one session copy" wait rather than relying on reader scarcity.

**MINOR**
- **#3 chunk size not validated** — fixed. A client-supplied `chunkFileSizeInBytes <= 0` now falls back to the configured default (`exportFileChunkSizeInBytes()`), same as an absent value, on both the export and start gRPC paths.
- **#4 unbounded `fileForFetchFuture().get()`** — fixed. Bounded to `get(1, TimeUnit.HOURS)`; on timeout the existing outer cleanup deletes the partial file and rethrows. (The background publish is not cancelled — that's the pre-existing behaviour noted in #8.)
- **#5 `sessionLocations.size()` O(n) pre-sizing** — **declined.** It's O(n) on a `ConcurrentLinkedDeque`, but trivially dominated by the O(n) walk that immediately follows, and the churn isn't worth it.

**NITS**
- **#6 `position += n` overflow** — fixed. `skip`/`skipNBytes` now reduce `n` modulo the buffer size before advancing, so a near-`Long.MAX_VALUE` `n` can't overflow `position` past the wrap guard.
- **#7 `mark()/reset()` position desync** — **left as-is** for now. It's dead code (no caller uses mark/reset) and the fix touches semantics; happy to address separately if you'd prefer `markSupported() → false`.
- **#8 `deleteFile` doesn't cancel the in-flight publish** — **declined** as out of scope: pre-existing behaviour shared with `TrafficRecorderTask`, not introduced here.

**Comment by novoj** [comment]
### Follow-up on this branch (post round-2 review)

**Deferred finding #7 — `RingBufferInputStream` mark/reset — now fixed** (`655e99c97`). `markSupported()` returns `false`, `reset()` throws, and `mark()` is a documented no-op. The previous overrides forwarded to the delegate file stream only, so a `reset()` restored the delegate's file pointer but left the ring-buffer `position` out of step — silently reading past the physical end of the ring window into unrelated bytes. It is dead code today (the only consumer wraps the stream in `ObservableInput`/Kryo and never calls mark/reset), so this just turns a latent silent-corruption trap into a clear "unsupported" signal.

**Merged current `dev` HEAD (`c1f9a6274`)** into the branch (`4b18ca706`) — clean, no conflicts.

**De-flaked two `dev`-authored tests** that came in with the merge and fail intermittently under CI load (`9b85dd708`). Both are test-side timing issues, not product regressions — the concurrency guard and the WAL/compaction machinery are correct. ⚠️ These two tests are owned by `dev`, not by this feature; this branch now carries the hardening.

- `ConcurrentSessionAccessTest.shouldRejectConcurrentAccessToReadWriteSession` — the writer thread re-inserted a **pre-seeded** primary key via `createNewEntity(pk).upsertVia()`, which carries `MUST_NOT_EXIST` semantics. Whenever the guard rejected the *reader* rather than the *writer*, the writer's insert executed and threw `InvalidMutationException` ("already entity PRODUCT with primary key 1 present") — unrelated to concurrency, both masking and racing the `ConcurrentSessionAccessException` the test asserts on. Reproduced at **8/9 failures under CPU load**; the writer now inserts a **unique** PK, verified **25/25 under the same load** (each pass also proves the guard still triggers).
- `EvitaTransactionalFunctionalTest.shouldRemoveOldDataFilesAndVerifyTimeTravel` — after polling up to 60 s for WAL cleanup, it slept a **fixed 250 ms** then asserted that background data-file compaction had removed the oldest catalog/collection files → "expected true but was false" under load. Now a bounded 60 s poll mirroring the WAL wait loop directly above it (4/4 under load).

Full `traffic_engine | grpc` suite green after the merge (501 tests, 0 failures, 2 pre-existing skips).
